### PR TITLE
:sparkles: Decompile java archives when given

### LINKED ISSUE #138: Decompile WAR files in java provider

### PR DESCRIPTION
**Pull request by djzager**
When the specified location for a java provider is to a `jar`, `war`, or `ear` we will unzip it and decompile any `.class` files found before running the analysis.

Fixes #138

### CODE REVIEW

**Comment by djzager** [comment]
If you want to play with this you will also need konveyor/java-analyzer-bundle#48

File: provider_container_settings.json [hunk 1]
```diff
@@ -11,7 +11,7 @@
         "name": "java",
         "binaryPath": "/jdtls/bin/jdtls",
         "initConfig": {
-            "location": "examples/java",
+            "location": "/analyzer-lsp/examples/java.jar",
```
Comment (djzager) [inline provider_container_settings.json:14]: This is for testing, should be pulled out into something else until we can specify multiple initConfigs. Might be appropriate to file an issue to follow-up on this once that's possible.

File: testing.Dockerfile [hunk 2]
```diff
@@ -5,7 +5,7 @@ COPY  . /analyzer-lsp
 RUN make build
 
 # The unofficial base image w/ jdtls and gopls installed
-FROM quay.io/konveyor/jdtls-server-base
+FROM jdtls-server-base
```
Comment (djzager) [inline testing.Dockerfile:8]: This should all be dropped...

File: testing.Dockerfile
Comment (djzager) [inline testing.Dockerfile:1]: This file shouldn't change.

File: rule-example.yaml
Comment (djzager) [inline rule-example.yaml:1]: I noticed an error 

```
java-pomxml-dependencies: 'open /analyzer-lsp/examples/java.jar/pom.xml: not a
      directory'
```

That we may need to handle. Additionally, I noticed some other things in the violation output that made me :thinking::

/analyzer-lsp/examples/java wasn't one of my providers, how'd this happen?

```
 - uri: file:///analyzer-lsp/examples/java/pom.xml
        message: POM XML dependencies
        variables:
          data: dependency
          innerText: "\n      io.fabric8\n      kubernetes-client-api\n      6.0.0\n
            \   "
          matchingXML: <groupId>io.fabric8</groupId><artifactId>kubernetes-client-api</artifactId><version>6.0.0</version>
```

We should look closely and make sure we are good.

File: provider/internal/java/util.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		}
+		defer archiveFile.Close()
+
+		if _, err := io.Copy(dstFile, archiveFile); err != nil {
+			return "", err
+		}
+
+		// If we found a class file, decompile it
+		if strings.HasSuffix(f.Name, ".class") {
+			cmd := exec.CommandContext(ctx, "java", "-jar", "/bin/fernflower.jar", filePath, path.Dir(filePath))
```
Comment (djzager) [inline provider/internal/java/util.go:76]: Might want to put fernflower.jar in `const`, or maybe wrap it in an executable shell script that can live in `/bin` so we don't need full path?

File: provider/internal/java/util.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	"path/filepath"
+	"strings"
+
+	"github.com/go-logr/logr"
+)
+
+// decompileJava is a function that extracts the contents ofa Java archive (.jar|.war|.ear) and
+// decompiles any .class files found using the fernflower decompiler and returns the path to
+// the destition directory on success, error otherwise.
+func decompileJava(ctx context.Context, log logr.Logger, location string) (string, error) {
```
Comment (djzager) [inline provider/internal/java/util.go:19]: Was hoping this would make it reusable later.

**Comment by djzager** [comment]
This should not be merged until https://github.com/konveyor/java-analyzer-bundle/pull/48 is merged.

File: docs/providers.md [hunk 5]
```diff
@@ -38,10 +38,26 @@ The `java` provider takes following additional configuration options:
 
 * `bundles`: Path to extension bundles to enhance default Java language server's capabilities. See the [bundle](https://github.com/konveyor/java-analyzer-bundle) Konveyor uses.
 * `workspace`: Path to directory where the provider generates debug information such as logs.
+* `location`: Path to the source code of the application to analyze. The location can be a binary jar, war, or ear file.
+
+Here's an updated example config for the Java provider in `provider_settings.json`:
+
+```json
```
Comment (shawn-hurley) [inline docs/providers.md:45]: this is changed so I wonder if we need to update this for the new settings.

Also changing again w/ the multi-init config change. do you think that we can get this in first and then do the API review and I can work on the rebase

**Review by shawn-hurley — APPROVED** [review]

**Pull request merged**